### DISCUSSION TITLE
Make it possible to view raw RLS policy SQL in the Supabase Dashboard

### DISCUSSION THREAD

**Discussion by TomasHubelbauer** [supabase/supabase#17330] [Feature Requests]
Hi!
I like that the Dashboard has a UI for creating policies from templates and that flow even previews the SQL that will be run to make (or amend) the policy, which is great!
But AFAICT it is not possible to view the SQL of the policy in the Dashboard once created? Barring the SQL editor of course where once could run a command to view all policies.
I wonder if a UI like this could be possible? It is the existing RLS policy section UI element but with an extra code block for the policy's SQL. Ideally with code highlighting of course, I skipped that for the purposes of this mock-up.
![image](https://github.com/supabase/supabase/assets/6831144/3778aef8-0921-4d44-899c-afa3e13f26ce)

I searched for "rls sql" but did not find an existing feature request.

**Comment by saltcod** [comment]
Great one @TomasHubelbauer.

**Comment by ThingEngineer** [comment]
A place, via Dashboard or CLI, to view a table's current RLS Policy(ies) SQL without having to diff changes and pull a new migration file would be a nice feature. Just add a code block tab for the policy's original SQL to the policy editor window. I don't care much for the policy editor window either. It's nice for simple stuff but to "wizardy" for my taste the majority of the time.

This is all you can recover from Postgres regarding the table policy so the sql would need a storage end point similar to how the sql editor stores past queries (on the hosted version).
```
SELECT 
  nspname AS schemaname,
  relname AS tablename,
  pol.polname AS policyname,
  pol.polpermissive,
  pg_get_userbyid(pol.polroles[1]) AS policyrole,
  pol.polcmd,
  pg_get_expr(pol.polqual, pol.polrelid) AS policyqual,
  pg_get_expr(pol.polwithcheck, pol.polrelid) AS policywithcheck
FROM 
  pg_policy pol
JOIN 
  pg_class cls ON pol.polrelid = cls.oid
JOIN 
  pg_namespace nsp ON cls.relnamespace = nsp.oid;
```

For a specific table just add on:

```
WHERE
  cls.relname = 'my_table_name';
```

**Reply by saltcod** [reply]
Thanks @ThingEngineer 

You can TAB here to get a copy of the current policy. Does this help? 

![screenshot-2024-02-13-at-15 55 55](https://github.com/supabase/supabase/assets/105593/80872cf2-7797-4202-9e18-7a47e8aa468e)

**Comment by rookledookle** [comment]
Upvote!!!!!

**Comment by joshdance** [comment]
Wait so rookie question, after I create a RLS policy using SQL, I can't view the SQL that created that policy?

**Comment by DiogoS** [comment]
I'm sorry, but this makes no sense at all.
So after I create a rule I can no longer view the raw sql?

**Reply by joshdance** [reply]
Yes in the Supabase interface, you can't see the raw SQL. 

This SQL command will output it, which you can run in their SQL editor on your own. 

But feels like it should be visible in the UI interface as well. 

```
SELECT
    -- Start building the CREATE POLICY statement string
    'CREATE POLICY "' || pol.polname || '"' ||  -- Policy name (quoted)
    ' ON ' || nsp.nspname || '.' || rel.relname || -- Table name (schema.table)
    CASE
        WHEN pol.polpermissive = true THEN ' AS PERMISSIVE'
        ELSE ' AS RESTRICTIVE' -- Although PERMISSIVE is default, being explicit is clear
    END ||
    ' FOR ' || CASE pol.polcmd -- Command type
                WHEN 'r' THEN 'SELECT'
                WHEN 'a' THEN 'INSERT'
                WHEN 'w' THEN 'UPDATE'
                WHEN 'd' THEN 'DELETE'
                WHEN '*' THEN 'ALL' -- Should rarely happen for specific policies, but covers it
                ELSE pol.polcmd::text -- Fallback just in case
            END ||
    -- Determine the TO clause. In Supabase, policies are often for 'authenticated'.
    -- The OID 16479 commonly corresponds to the 'authenticator' role, which the
    -- 'authenticated' pseudo-role effectively executes as. If polroles is null
    -- or empty, it means TO public.
    CASE
        WHEN pol.polroles IS NULL OR array_length(pol.polroles, 1) = 0 THEN ' TO public'
        -- Based on your CSV, [16479] seems to map to 'TO authenticated'.
        -- If you had policies for other specific roles by name, this logic would
        -- need to be more complex (joining pg_roles and aggregating names).
        -- For the common Supabase case with [16479], 'TO authenticated' is appropriate.
        ELSE ' TO authenticated'
    END ||
    -- Add the USING clause if the expression exists
    CASE
        WHEN pg_get_expr(pol.polqual, pol.polrelid) IS NOT NULL THEN ' USING (' || pg_get_expr(pol.polqual, pol.polrelid) || ')'
        ELSE '' -- No USING clause
    END ||
    -- Add the WITH CHECK clause if the expression exists
    CASE
        WHEN pg_get_expr(pol.polwithcheck, pol.polrelid) IS NOT NULL THEN ' WITH CHECK (' || pg_get_expr(pol.polwithcheck, pol.polrelid) || ')'
        ELSE '' -- No WITH CHECK clause
    END ||
    ';' -- Add a semicolon to terminate the statement
AS create_policy_sql -- Alias the generated string column
FROM
    pg_catalog.pg_policy pol
JOIN
    pg_catalog.pg_class rel ON pol.polrelid = rel.oid
JOIN
    pg_catalog.pg_namespace nsp ON rel.relnamespace = nsp.oid
WHERE
    nsp.nspname = 'public'
    AND rel.relname = 'PutYourTableNameHere';
```
